### ISSUE TITLE
Agent ServicePodWorker re-applies after any apply that triggers container events

### ISSUE THREAD

**Issue by SpComb** [kontena/kontena#2766]
Any Docker container actions done in `ServicePodWorker#apply` will also result `container:event` notifications. Because the `apply` function runs as an `Celluloid.exclusive { ... }` block, those notifications get queued up during the deploy, and get run once the `apply` returns:

```
D, [2017-09-05T12:50:35.514173 #1] DEBUG -- Kontena::RpcClient: waited 0.2s of 30.0s until: request /node_service_pods/list has response wth id=1165729746 yielded Array
D, [2017-09-05T12:50:35.515733 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of log-test-1: running
I, [2017-09-05T12:50:35.518770 #1]  INFO -- Kontena::Workers::ServicePodWorker: creating log-test-1
...
I, [2017-09-05T12:50:37.949590 #1]  INFO -- Kontena::ServicePods::Creator: service started: log-test-1
D, [2017-09-05T12:50:38.085292 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event create
D, [2017-09-05T12:50:38.088439 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event start
```

With the way that the `@container_state_changed` works, that means that each `update` -> `apply if needs_apply?` essentially runs twice: once because the `@service_pod` changes, and then a second time because of the `@container_state_changed = true` triggerd by the `apply` itself:

```
D, [2017-09-05T12:50:48.685974 #1] DEBUG -- Kontena::RpcClient: waited 0.0s of 30.0s until: request /node_service_pods/list has response wth id=1628267283 yielded Array
D, [2017-09-05T12:50:48.948312 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of log-test-1: running
D, [2017-09-05T12:50:49.086853 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state is in-sync: running
```

The second `apply` is a no-op because the container is up to date: https://github.com/kontena/kontena/pull/2143#discussion_r111080598

**Comment by SpComb** [comment]
This is now particularly apparent with #2689 during a deploy that re-creates the container:

```
D, [2017-09-05T13:00:41.232281 #1] DEBUG -- Kontena::RpcServer: rpc request: Kontena::Rpc::ServicePodsApi#notify_update [[]]
D, [2017-09-05T13:00:41.275623 #1] DEBUG -- Kontena::RpcClient: waited 0.0s of 30.0s until: request /node_service_pods/list has response wth id=173536812 yielded Array
D, [2017-09-05T13:00:41.527638 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of log-test-1: running
D, [2017-09-05T13:00:41.550882 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container before ensure: fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b
I, [2017-09-05T13:00:41.553790 #1]  INFO -- Kontena::Workers::ServicePodWorker: service updated at 2017-09-05T13:00:41+00:00 after service container created at 2017-09-05T12:59:51+00:00
I, [2017-09-05T13:00:41.557358 #1]  INFO -- Kontena::Workers::ServicePodWorker: re-creating log-test-1
I, [2017-09-05T13:00:43.540437 #1]  INFO -- Kontena::ServicePods::Creator: removing previous version of service: log-test-1
D, [2017-09-05T13:00:54.086130 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: log-test-1
D, [2017-09-05T13:00:54.716991 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container after ensure: 6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c
```

Followed by:

```
D, [2017-09-05T13:00:54.769163 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b status=kill
D, [2017-09-05T13:00:54.933915 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b status=kill
D, [2017-09-05T13:00:54.935607 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b status=die
I, [2017-09-05T13:00:54.936627 #1]  INFO -- Kontena::Workers::ServicePodWorker: restarting log-test-1 because it has stopped
D, [2017-09-05T13:00:54.937170 #1] DEBUG -- Kontena::Workers::ServicePodWorker: restart triggered (from 2017-09-05 13:00:54 UTC)
D, [2017-09-05T13:00:54.937278 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of log-test-1: running
D, [2017-09-05T13:00:54.944830 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container before ensure: 6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c
D, [2017-09-05T13:00:55.027842 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state is in-sync: running
D, [2017-09-05T13:00:55.028021 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container after ensure: 6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c
D, [2017-09-05T13:00:55.038402 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b status=stop
D, [2017-09-05T13:00:55.038667 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=fe8e0bf69c46075583d3b39508aff87407c553eb8734d76a40d98efafbf34d3b status=destroy
D, [2017-09-05T13:00:55.038831 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c status=create
D, [2017-09-05T13:00:55.038967 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event id=6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c status=start
```

And then also

```
D, [2017-09-05T13:00:56.742337 #1] DEBUG -- Kontena::RpcClient: waited 0.1s of 30.0s until: request /node_service_pods/list has response wth id=1401174148 yielded Array
D, [2017-09-05T13:00:57.007340 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of log-test-1: running
D, [2017-09-05T13:00:57.021707 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container before ensure: 6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c
I, [2017-09-05T13:00:57.036965 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image: redis:latest
I, [2017-09-05T13:00:58.806973 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: redis:latest
D, [2017-09-05T13:00:58.827775 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state is in-sync: running
D, [2017-09-05T13:00:58.827831 #1] DEBUG -- Kontena::Workers::ServicePodWorker: service log-test-1 container after ensure: 6267583e6a4eb9695ca1070bd4f980471723d47b01590536d5fd7138ca88b85c
```

The `restarting log-test-1 because it has stopped` gets logged after every `kontena service deploy --force` now.

**Comment by SpComb** [comment]
This isn't 100% fixed, it ignores the `kill`, `die`, `stop`, `destroy` events, but it still picks up the `create` and `start` events and re-applies the same `deploy_rev` again. There could still be some potential for simplifying the actor state by removing the `@container_state_changed`, assuming that triggering the restart logic on `die` events is enough?

**Comment by SpComb** [comment]
Re-opening this to cover not just the `restart on die` thing fixed in #2773, but also the `apply on start` thing that makes the concurrent `wait_for_port` in #2776 so much more confusing
